DOC update model persistence page to provide clearer recommendations

rprkh opened the pull request:
#### Reference Issues/PRs
Fixes #27514

#### What does this implement/fix? Explain your changes.
Updates the model persistence page and provides more information on the different approaches used to persist models.

#### Any other comments?
Based on https://github.com/scikit-learn/scikit-learn/issues/27514#issue-1920970883, the tasks to be completed as part of this PR are as follows:

- [x] List all alternatives near the beginning of the document.
- [x] Start by discussing the fact that you need to pin all transitive scikit-learn dependencies to be able to safely load a pickle, which is the format most people will use, and probably gets you 80% of the benefit you will get from more complicated recommendations.
- [x] Do more to compare the presented alternatives (what does skops lack that would make one want to use ONNX?)
- [x] Summarize the most important points at the end of the document.

rprkh committed:
make model persistence doc clearer

rprkh committed:
render links

rprkh committed:
update onnx description

rprkh committed:
minor improvement

rprkh committed:
spelling

glemaitre reviewed (commented):
I started to have a look and I think that we need more changes regarding the structure if we want to address the original issue.

We will have to do a major rework of the page sectioning. For instance, instead of starting with "Python Specific Serialization" I think that we need to have first section on the principle behind the process of serializing and loading the artifact.

Indeed, I would expect the first section to only present the workflow. We can illustrate with `pickle` (or maybe `skop` or `joblib` as a default). In summary I would expect something like:

- Store the model in an artifact that we can later open
- On the production environment, create a similar environment from the one used to store the artifact
- Load the artifact
- Serve it

In between these steps, we can have the discussion about the versioning and freezing and mentioning some solution out in the pydata ecosystem.

Then, we can go into details with the different methods to store/load models and discuss the concerns:

- Python pickle: not optimum for large numpy arrays (this would need to be checked since some work in the latest pickle protocol has been done but @ogrisel knows better)
- joblib alternative for a better handling of arrays
- skops to load only pickle only from trusted modules
- onnx and sklearn-onxx when we are only interested the inference part of the model (it is sometimes more efficient) and the runtime is then independent from the environment used to train.
- I don't know enough about PMML

rprkh committed:
add suggestions

rprkh commented:
> Indeed, I would expect the first section to only present the workflow. We can illustrate with pickle (or maybe skop or joblib as a default). In summary I would expect something like:
> 
> Store the model in an artifact that we can later open
On the production environment, create a similar environment from the one used to store the artifact
Load the artifact
Serve it

I have illustrated this section with a simple example on the usage of `pickle`.

> In between these steps, we can have the discussion about the versioning and freezing and mentioning some solution out in the pydata ecosystem.

I have added this to the "Replicating the training environment in production" subsection.

> Then, we can go into details with the different methods to store/load models and discuss the concerns:
>
> Python pickle: not optimum for large numpy arrays (this would need to be checked since some work in the latest pickle protocol has been done but @ogrisel knows better)
joblib alternative for a better handling of arrays
skops to load only pickle only from trusted modules
onnx and sklearn-onxx when we are only interested the inference part of the model (it is sometimes more efficient) and the runtime is then independent from the environment used to train.
I don't know enough about PMML

I have added some of these points to the "Summarizing the keypoints" section. It is a concise version of the pros/cons of each approach, and describes the most important points in approximately 1 line each.

rprkh committed:
update note

rprkh committed:
link

rprkh committed:
fix

ArturoAmorQ commented on doc/model_persistence.rst:
```diff
@@ -182,3 +270,69 @@ not help in production when performance is critical.
 <https://github.com/jpmml/sklearn2pmml>`_ distributed under the Affero GPLv3
 license.
+
+Choosing the right approach for model persistence
+-------------------------------------------------
```
I have the sensation this whole section says almost the same as the `Summarizing the keypoints` section afterwards. I would rather keep the latter for easier reading, but maybe enrich the summary if needed, e.g.
```
* `ONNX`: It provides a uniform format for persisting any machine learning
  or deep learning model (other than scikit-learn) and is useful
  for model inference. It can however, result in compatibility issues with
  different frameworks.
```

ArturoAmorQ reviewed (commented):
Thanks for the PR @rprkh! It's looking in good shape :) Here is a batch of comments.

rprkh committed:
include new suggestions

rprkh committed:
Update doc/model_persistence.rst

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

rprkh committed:
Merge branch 'main' into doc_model_persistance

rprkh commented:
Hey @ArturoAmorQ, I have added the suggestions.

ArturoAmorQ commented on doc/model_persistence.rst:
```diff
@@ -181,3 +267,29 @@ not help in production when performance is critical.
+  While `pickle` can be used to easily save and load scikit-learn models,
+  unpickling of untrusted data might lead to security issues.
+* `joblib`: Efficient storage and memory mapping techniques make it faster
+  for serializing and deserializing large numpy arrays. However, care must be
+  taken when loading untrusted data as it can lead to the execution of
+  arbitrary Python code, which may be malicious.
```
```suggestion
  when working with large machine learning models or large numpy arrays. However,
  it may trigger the execution of malicious code while loading untrusted data.
```

ArturoAmorQ reviewed (approved):
Just a tweak but otherwise LGTM :)

@ogrisel or @glemaitre, would you like to have a word on this?

adrinjalali committed:
Update doc/model_persistence.rst

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

adrinjalali reviewed (approved):
This is an improvement, let's merge, and I'll open a successive PR with more improvements.
